core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4

```diff
@@ -472,6 +534,43 @@ frameBound
     | expression boundType=(PRECEDING | FOLLOWING)  #boundedFrame
     ;
 
+patternAlternation
```

Comment:
It may be possible to flatten all these (patternAlternation, patternConcatenation, quantifiedPrimary) into a single rule, similar to valueExpression. The order of precedence is determined by the order of the productions

Try something like:

```
pattern
    :  identifier
    | '(' ')'
    | PERMUTE '(' pattern (',' pattern)* ')'
    | '(' pattern ')'
    | '^'
    | '$'
    | '{-' pattern '-}'
    | pattern patternQuantifier?
    | pattern pattern
    | pattern ('|' pattern)*
    ;
```

Reply:
It was possible to simplify it to some extent, though not to a single rule.
After I simplified the grammar, the produced structures are nested where they used to be flat, E.g. `A | B | C` used to parse as `PatternAlternation(A, B, C)`. Now it is `PatternAlternation(PatternAlternation(A, B), C)`. 
If we want the grammar simplified, we should consider adding a flattening transformation for the resulting patterns.